### PR TITLE
feat(kdf.js): Introduce key derivation functions

### PR DESCRIPTION
**Pull request by LesterKim**
Fixes #

## Proposed Changes

  -
  -
  -

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -379,6 +379,38 @@ const server = kx.server({
...
 ```
 
+### `crypto.kdf.keygen(key)` <a name="kdf-keygen"></a>
+
+> **Stability: 2** - Stable
+
+Generate a key of length `crypto_kdf_KEYBYTES`.
+
+```js
+const { crypto_kdf_KEYBYTES } = require('ara-crypto/sodium')
```
Comment (jwerle) [inline README.md]: The user of this module should not need to interact with sodium directly. the `keygen` function here should be functional. That is, the function allocates the buffer for the caller, and returns it.

Comment (jwerle) [reply README.md]: Could you advertise the functional approach for key generation?

File: kdf.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+ *
+ * @public
+ * @param {?(Buffer)} [subkey]
+ * @param {?(Number)} [subkeyId]
+ * @param {Buffer} ctx
+ * @param {Buffer} key
+ * @return {Buffer}
+ * @throws TypeError
+ */
+function derive(subkey, subkeyId, ctx, key) {
```
Comment (jwerle) [inline kdf.js]: Perhaps this function should be a bit lower level and a higher level one could exist that simply allows one to give a `key` from `kdf.keygen()` and pass it to a `kdf.derive(key, iterations)` that returns the final subkey. I imagine the `derive` function you currently have being broken up into a `ctx = kdf.init(key)`, `subkey kdf.update(ctx, id)`, and `kdf.final(ctx)` which `kdf.derive()` should compose. A simple an unchecked example can be seen below:

```js
function init(key) {
  return { key, buffer: Buffer.allocUnsafe(crypto_kdf_CONTEXTBYTES), subkey: null }
}

function update(ctx, id) {
  ctx.subkey = ctx.subkey || Buffer.allocUnsafe(crypto_kdf_KEYBYTES)
  crypto_kdf_derive_from_key(ctx.subkey, id, ctx.buffer, ctx.key)
  return ctx.subkey
}

function final(ctx) {
  const subkey = ctx.subkey
  ctx.subkey = null
  return ctx.subkey
}

function derive(key, iterations) {
  const ctx = init(key)
  for (let i = 0; i < iterations.length; ++i) {
    update(ctx, i) // or i + i
  }
  return final(ctx)
}
```

File: README.md (hunk 1 above)
Comment (LesterKim) [reply README.md]: Yes, I'll make the change.

File: kdf.js (hunk 2 above)
Comment (LesterKim) [reply kdf.js]: Ok, I'll break it down into smaller parts.

Comment (jwerle) [reply kdf.js]: Well, `init()` could be rewritten to account for a given context buffer and already initialized `subkey`:

```js
function init(key, buffer, subkey)
  return {
    key,
    buffer: buffer || Buffer.allocUnsafe(crypto_kdf_CONTEXTBYTES),
    subkey: subkey || null }
}
```

Comment (LesterKim) [reply kdf.js]: @jwerle Do you want me to disregard stylistically how all the other files do error handling or would you prefer I stay consistent while incorporating your suggestions?

Comment (jwerle) [reply kdf.js]: The latter please. Consistent is better 🙃

Comment (LesterKim) [reply kdf.js]: 👍

File: kdf.js [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+/**
+ * Initializes key derivation.
+ *
+ * @public
+ * @param {Buffer} buffer
+ * @param {Buffer} key
+ * @return {Object}
+ * @throws TypeError
+ */
+function init(buffer, key) {
```
Comment (jwerle) [inline kdf.js]: We should not need to supply a buffer. This function requires the caller to supply a buffer or a falsy value like `null` to get the "default" effect. The optional buffer should be the second argument.

File: kdf.js [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+ * Derives a subkey using the master key and context.
+ *
+ * @public
+ * @param {Buffer} buffer
+ * @param {Buffer} key
+ * @param {Number} iterations
+ * @return {Buffer}
+ * @throws TypeError
+ */
+function derive(buffer, key, iterations) {
```
Comment (jwerle) [inline kdf.js]: This function requires the caller to supply a buffer or falsy argument. The caller should only have to supply the key and iterations and potentially, an optional buffer, which should be in the last argument position.

File: README.md [hunk 5]
```diff
@@ -379,6 +379,34 @@ const server = kx.server({
...
+
+> **Stability: 2** - Stable
+
+Generate a master ("secret") key.
+
+```js
+const key = crypto.kdf.keygen()
+```
+
+### `crypto.kdf.derive(subkey, subkeyId, ctx, key)` <a name="kdf-derive"></a>
```
Comment (jwerle) [inline README.md]: Can you correct the documentation too?

**Comment by LesterKim** [comment]
@vipyne Good call. Adding now

**Pull request merged**